File: clients/console/views/transport.js
/*
    This file is part of Go.dot — https://github.com/pob31/go.dot

    Copyright (C) 2026 Pierre-Olivier Boulant

    Go.dot is free software: you can redistribute it and/or modify it under the
    terms of the GNU General Public License as published by the Free Software
    Foundation, either version 3 of the License, or (at your option) any later
    version. Go.dot is distributed in the hope that it will be useful, but
    WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
    or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License
    (LICENSE, at the repository root) for more details.

    SPDX-License-Identifier: GPL-3.0-or-later
*/

/*  THE TRANSPORT'S BUTTONS: GO, the lock, revert, and every button that names
    its command in its own `data-cmd`. Wired when this module is loaded. */

import { tree } from "../plumbing/tree.js";
import { command, send } from "../plumbing/link.js";
import { gesture } from "../gestures/table.js";
import { wantsTheKeys } from "../gestures/keys.js";
import { el } from "./common.js";

el("go").addEventListener("click", () => gesture("go", []));

/*  WHERE THE KEYBOARD IS, IN WORDS.

    A box with the focus keeps the arrows and the accelerators - stepping a
    number is what a number box is for - so while one has them the hint beside
    GO would otherwise be telling an operator that ctrl-Z undoes, when ctrl-Z
    is the browser's typing-undo and the arrows are moving a pre-wait rather
    than the standby. Didi's time columns put three such boxes on every row, so
    this is now one click away at all times.

    SAID, AND NOT COLOURED (§4.8). The focus ring on a box is a pixel of hue on
    a control the pointer is already resting on; the sentence is what an
    operator reads. Space is still GO from a number box (gestures/keys.js), and
    the sentence says that too, because the one thing nobody may be made to
    wonder about is whether GO will fire.

    ON THE DOCUMENT, because focus moves between elements this module does not
    own; `focusout` runs before the next element has it, so the answer is asked
    for after the browser has settled. */
const hint = el("hint");
const HINT_KEYS = hint ? hint.textContent : "";
const HINT_TYPING = "typing in a box · space still GO · esc gives the keyboard back";

function sayWhereTheKeysAre() {
  if (! hint) return;

  const typing = wantsTheKeys(document.activeElement);

  hint.dataset.typing = typing ? "yes" : "no";

  const said = typing ? HINT_TYPING : HINT_KEYS;

  if (hint.textContent !== said) hint.textContent = said;
}

document.addEventListener("focusin", sayWhereTheKeysAre);
document.addEventListener("focusout", () => setTimeout(sayWhereTheKeysAre, 0));

/*  THE LOCK IS A NODE, NOT A COMMAND (namespace draft §14.7): the toggle is a
    `node.set` on /godot/document/locked like any other write, carrying the
    boolean it wants rather than a word, and the engine's answer comes back the
    way every answer does - as the reading the next poll brings. Which way to
    flip is decided from what the engine last said, not from what this button
    last did, so two pages in two hands cannot argue it into the wrong state. */
el("lock").addEventListener("click", () => {
  const now = tree.get("/godot/document/locked", null);
  const isLocked = now === true || now === "true";

  send("/godot/document/locked", [{ tag: isLocked ? "F" : "T" }]);
});

/*  REVERT ASKS FIRST AND THE DELETE DOES NOT, and that is not an inconsistency.
    PR 5.4 took the confirm off `object.delete` because undo had arrived: a
    gesture ctrl/⌘-Z takes back does not need a dialogue, and a dialogue on a
    gesture that can be taken back teaches an operator to dismiss dialogues.
    Revert is the case that argument does not reach. It reloads the show from
    disk and CLEARS THE UNDO HISTORY as part of the same command (§14.10), so
    afterwards there is nothing on the stack to press ctrl-Z on - and what goes
    is not one edit but everything since the last save, in one press, with
    nowhere left to fetch it from. Undo is the whole of why the delete stopped
    asking, and undo is exactly what revert takes away, so this is where the
    page asks. The question says what will be lost rather than asking whether
    anybody is sure, because "are you sure?" is a question nobody reads.

    Discard, in the recovery banner, is as final and does not ask twice: the
    banner it sits in is already the question, written out in full, and a
    dialogue over a paragraph somebody has just read is that question again. */
el("revert").addEventListener("click", () => {
  const show = tree.get("/godot/document/name", "") || "the show";

  if (window.confirm("Revert " + show + " to what was last saved?\n\n" +
                     "Everything changed since then is thrown away, and undo cannot " +
                     "bring it back: reverting clears the undo history as well.")) {
    gesture("revert", []);
  }
});

for (const button of document.querySelectorAll("#transport [data-cmd], #recovery [data-cmd], #pane-gogo header [data-cmd]")) {
  button.addEventListener("click", () => command(button.dataset.cmd));
}

File: src/wfg/client/model/Transport.cpp
/*
    This file is part of Go.dot — https://github.com/pob31/go.dot

    Copyright (C) 2026 Pierre-Olivier Boulant

    Go.dot is free software: you can redistribute it and/or modify it under the
    terms of the GNU General Public License as published by the Free Software
    Foundation, either version 3 of the License, or (at your option) any later
    version. Go.dot is distributed in the hope that it will be useful, but
    WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
    or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License
    (LICENSE, at the repository root) for more details.

    SPDX-License-Identifier: GPL-3.0-or-later
*/

#include <wfg/client/model/Transport.h>

#include <wfg/engine/tree/TreeSnapshot.h>

#include <cstddef>
#include <cstdint>
#include <string>
#include <string_view>
#include <tuple>

namespace wfg::client::model
{
    namespace
    {
        /*  THE DASH IS "THE ENGINE HAS NOT SAID", and it is the same dash the
            page draws, for the same reason: a reading nobody has published yet
            is not a zero and not a no. */
        constexpr const char* unsaid = "—";

        /*  `join`, which put a middle dot between two half-sentences, went with
            the last line that had two halves to put together: the transport
            lost its saved/unsaved line and its audio word on 2026-09-18, and
            what is left says one thing each. Kept out rather than kept unused -
            GCC's -Werror=unused-function is right that a helper nothing calls
            is a claim about a shape the file no longer has. */
    }

    std::string TransportReading::standbyLine() const
    {
        if (standbyId.empty())
            return "no standby";

        if (standbyKind.empty())
            return standbyName;

        return standbyName + "  " + standbyKind;
    }

    namespace
    {
        /*  ONE HALF OF THE OLD UNDO LINE, now a tooltip apiece. `—` for a
            client the engine has not answered yet, the plain refusal when
            there is nothing, and otherwise the engine's own name for what
            would be taken back. */
        std::string tipFor (Flag can, const std::string& name,
                            const char* verb, const char* nothing)
        {
            if (can == Flag::unsaid)    return std::string (verb) + ": " + unsaid;
            if (can == Flag::no)        return std::string (nothing);

            return std::string (verb) + " " + (name.empty() ? "the last edit" : name);
        }
    }

    std::string TransportReading::undoTip() const
    {
        return tipFor (canUndo, undoName, "undo", "nothing to undo");
    }

    std::string TransportReading::redoTip() const
    {
        return tipFor (canRedo, redoName, "redo", "nothing to redo");
    }

    std::size_t countWarnings (std::string_view all)
    {
        if (all.empty())
            return 0;

        /*  ONE PER LINE, so the count is the lines - and a trailing newline
            does not invent a last empty warning. */
        auto lines = std::size_t { 1 };

        for (std::size_t at = 0; at < all.size(); ++at)
            if (all[at] == '\n' && at + 1 < all.size())
                ++lines;

        return lines;
    }

    std::string firstWarning (std::string_view all)
    {
        const auto end = all.find ('\n');
        const auto first = all.substr (0, end == std::string_view::npos ? all.size() : end);

        /*  CLIPPED HERE, not by whatever draws it. One warning of a quarter of
            a megabyte is as able to hang a text layout as eighteen hundred
            short ones, and the place that knows this is a foot-of-window
            summary is here. */
        constexpr std::size_t longest = 160;

        if (first.size() > longest)
            return std::string (first.substr (0, longest)) + "…";

        return std::string (first);
    }

    std::string TransportReading::warningLine() const
    {
        if (warningCount == 0)
            return {};

        const auto count = std::to_string (warningCount)
                         + (warningCount == 1 ? " warning" : " warnings");

        return warningFirst.empty() ? count : count + " · " + warningFirst;
    }

    std::string TransportReading::errorLine() const
    {
        if (status == "noClock")
            return "Audio disconnected - cues paused; waiting for the interface and clock.";
        if (lastError.empty())
            return {};

        /*  tick, sequence, origin, reason, command - and the command may carry
            no spaces, so five fields is exactly what a well-formed record has.
            Anything else is shown as it came. */
        const auto fields = words (lastError);

        if (fields.size() != 5)
            return lastError;

        return fields[4] + " refused: " + fields[3];
    }

    std::string TransportReading::lockLine() const
    {
        /*  `unsaid` gets no word at all, because "the engine has not told us
            whether the show is locked" belongs beside the lock's own button,
            which says so by being disabled. */
        return locked == Flag::yes ? "locked" : "";
    }

    bool TransportReading::operator== (const TransportReading& other) const noexcept
    {
        const auto tie = [] (const TransportReading& r)
        {
            return std::tie (r.show, r.dirty, r.locked, r.recovery, r.recording,
                             r.tick, r.clock, r.rate,
                             r.listId, r.listName, r.standbyId, r.standbyName, r.standbyKind,
                             r.standbyNotes,
                             r.canUndo, r.canRedo, r.undoName, r.redoName,
                             r.status, r.lastError, r.writeError,
                             r.warningCount, r.warningFirst, r.revision);
        };

        return tie (*this) == tie (other);
    }

    TransportReading readTransport (const tree::TreeSnapshot& snapshot)
    {
        TransportReading reading;

        reading.show = text (snapshot, "/godot/document/name");
        reading.dirty = flag (snapshot, "/godot/document/dirty");
        reading.locked = flag (snapshot, "/godot/document/locked");
        reading.recovery = flag (snapshot, "/godot/document/recovery");
        reading.recording = flag (snapshot, "/godot/document/recording");

        reading.tick = text (snapshot, "/godot/engine/tick");
        reading.clock = text (snapshot, "/godot/engine/clock");

        const auto rate = text (snapshot, "/godot/engine/sampleRate");
        const auto block = text (snapshot, "/godot/engine/blockSize");
        reading.rate = (rate.empty() || rate == "0") ? std::string() : rate + " / " + block;

        /*  THE FOCUSED LIST, WITH THE PAGE'S FALLBACK (strip.js:223): the
            engine's `focus` when it names one, else the first list there is,
            else nothing - a show with no list has no standby to show. */
        reading.listId = text (snapshot, "/godot/list/focus");

        if (reading.listId.empty())
            if (const auto lists = words (text (snapshot, "/godot/list/order")); ! lists.empty())
                reading.listId = lists.front();

        if (! reading.listId.empty())
        {
            reading.listName = text (snapshot, "/godot/list/" + reading.listId + "/name");
            reading.standbyId = text (snapshot, "/godot/list/" + reading.listId + "/standby");
        }

        if (! reading.standbyId.empty())
        {
            reading.standbyName = text (snapshot, "/godot/cue/" + reading.standbyId + "/name");
            reading.standbyKind = text (snapshot, "/godot/cue/" + reading.standbyId + "/kind");
            reading.standbyNotes = text (snapshot, "/godot/cue/" + reading.standbyId + "/notes");
        }

        reading.canUndo = flag (snapshot, "/godot/document/canUndo");
        reading.canRedo = flag (snapshot, "/godot/document/canRedo");
        reading.undoName = text (snapshot, "/godot/document/undoName");
        reading.redoName = text (snapshot, "/godot/document/redoName");

        reading.status = text (snapshot, "/godot/audio/status");
        reading.lastError = text (snapshot, "/godot/engine/lastError");
        reading.writeError = text (snapshot, "/godot/document/writeError");
        /*  READ, SUMMARISED, AND THE LONG STRING DROPPED on the spot: nothing
            downstream of here ever holds it, so nothing downstream can be hung
            by a show with eighteen hundred things wrong with it. */
        const auto warnings = text (snapshot, "/godot/document/warnings");
        reading.warningCount = countWarnings (warnings);
        reading.warningFirst = firstWarning (warnings);

        if (const auto* node = snapshot.find ("/godot/document/revision"))
            if (const auto sole = node->soleValue(); sole.has_value() && sole->isInt64())
                reading.revision = static_cast<std::uint64_t> (sole->getInt64());

        return reading;
    }
}
